platform/server/http/Router/Request.ts

```diff
@@ -0,0 +1,83 @@
+import * as express from 'express';
+import { schema } from '@elastic/kbn-utils';
+
+import { Headers, filterHeaders } from './headers';
+import { RouteSchemas } from './Route';
+
+export class KibanaRequest<Params, Query, Body> {
+  readonly headers: Headers;
+
+  /**
+   * Factory for creating requests. Validates the request before creating an
+   * instance of a KibanaRequest.
+   */
+  static from<
+    P extends schema.ObjectSetting,
+    Q extends schema.ObjectSetting,
+    B extends schema.ObjectSetting
+  >(req: express.Request, routeSchemas: RouteSchemas<P, Q, B> | undefined) {
+    const requestParts = KibanaRequest.validate(req, routeSchemas);
+    return new KibanaRequest(
+      req,
+      requestParts.params,
+      requestParts.query,
+      requestParts.body
+    );
+  }
+
+  /**
+   * Validates the different parts of a request based on the schemas defined for
+   * the route. Builds up the actual params, query and body object that will be
+   * received in the route handler.
+   */
+  private static validate<
+    P extends schema.ObjectSetting,
+    Q extends schema.ObjectSetting,
+    B extends schema.ObjectSetting
+  >(
+    req: express.Request,
+    routeSchemas: RouteSchemas<P, Q, B> | undefined
+  ): {
+    params: schema.TypeOf<P>;
```

Comment:
`schema.TypeOf` basically means: "give me the type of the result of validating the schema", e.g. let's say we have this schema:

```
const test = schema.object({
  foo: schema.string()
});
```

When `validate` is called with schema in `params`,  it will set `P = schema.ObjectSetting<{ foo: schema.StringSetting }>` (which is in case matches the `extends schema.ObjectSetting`, so this is considered a valid type). We couldn't call this with just a string or an object, as those wouldn't `extends schema.ObjectSetting`.

So the final type of the `params` being returned here (aka the result of `schema.TypeOf<P>`) is:

```
{
  foo: String
}
```

aka an object with a key named `foo` of type `String`